### ISSUE TITLE
Waypoint user note: Out of sync-warning for variables

### ISSUE THREAD

**Issue by moving-bits** [cgeo/cgeo#13878]
_Originally posted by @moving-bits in https://github.com/cgeo/cgeo/issues/13867#issuecomment-1383122343_

Thanks for adding this nice feature. 

I played with it a bit (works nicely), but would like to see an additional feature:
If you have a variable in a waypoint user note, which is out of sync, and you close the user note, it would be great to get a notification (popup?) for this to be able to get vars back in sync.

- Added a waypoint to a trad. cache
- Opened this waypoint from map, entered A=1, B=3 into separate lines of user note => Variables got created, fine.
- Opened waypoint from map again, changed A to 5 => Variable in variables list got changed, fine.
- Changed value of A in variables view to 7, opened waypoint from map again, changed A to 6 => Variable in variables list did not change (ok), but you do not get a notification about vars being out of sync.

I understand that you do not want to change the user note text (IIRC), but on receiving a notification the user could decide between change value in cache-global variables list, change value in user note manually, or ignore the difference.

**Comment by moving-bits** [comment]
_Originally posted by @eddiemuc in https://github.com/cgeo/cgeo/issues/13867#issuecomment-1383125949_

Thanks for the feedback

> ...but would like to see an additional feature: If you have a variable in a waypoint user note, which is out of sync, and you close the user note, it would be great to get a notification (popup?) for this to be able to get vars back in sync.
> ...
> I understand that you do not want to change the user note text (IIRC), but on receiving a notification the user could decide between change value in cache-global variables list, change value in user note manually, or ignore the difference.

Potential problem I see here: when the user decides NOT to sync, then he/she will get this popup over and over again whenever the waypoint is edited. This could be a real pain. 

On the other hand, if we decide that note values should **always** override the var list then we don't need a popup at all.

I personally would not change the user note text, but this is a group decision not my personal one.

Would it be ok to handle this in a separate, new issue (not in this PR)?

**Comment by moving-bits** [comment]
_Originally posted by @moving-bits in https://github.com/cgeo/cgeo/issues/13867#issuecomment-1383127187_

> Potential problem I see here: when the user decides NOT to sync, then he/she will get this popup over and over again whenever the waypoint is edited. This could be a real pain.

understandably. So maybe just add an "out of sync" button (like you have done for cache note), if vars are out of sync. Less intrusive, and user can decide how to react. But I would place the button above the input field in this case, as it may be hidden below the keyboard otherwise.

> Would it be ok to handle this in a separate, new issue (not in this PR)?

sure

**Comment by moving-bits** [comment]
_Originally posted by @eddiemuc in https://github.com/cgeo/cgeo/issues/13867#issuecomment-1383131332_

> So maybe just add an "out of sync" button (like you have done for cache note), if vars are out of sync. Less intrusive, and user can decide how to react.

Thought about this myself. But there's a big difference in waypoint edit view when compared to cache personal note: the screen is the same in which the actual text-typing for the user note is happening. So unless we implement a "live check" function continuously checkng whether the just-typed-text would (if saved) create an asyncronity in var list (and this is quite complex), then the button would not appear "in-time" but only after the "edit waypoint" view is closed/saved and then reopened. Also the other way around: if there's an asynchronity and I correct it by changing the text, then button would stay unless i closed/save and reopen the waypoint edit view. This seemed too complex to me for the problem at hand, so I stayed with the button at cache user note.

**Comment by moving-bits** [comment]
Found a possible way to listen to changes in text field, for continuously checking button state:

```
binding.userNote.addTextChangedListener(new TextWatcher() {
    @Override
    public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
        // do nothing
    }

    @Override
    public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
        // do nothing
    }

    @Override
    public void afterTextChanged(Editable editable) {
        Log.e("EditText content: " + editable.toString());
    }
});
```

**Comment by eddiemuc** [comment]
> Found a possible way to listen to changes in text field, for continuously checking button state:

That's not the problematic part. I use this feature myself e.g. in variable tab for on-type-formula-evaluation (I even created a helper function to quickly create a TextWatcher, see `ViewUtils.createSimpleWatcher()`

The problematic point is that this will de facto lead to calling a rather complex parsing function each time a letter is typed - just for the purpose of enabling/disabling a button.  To prevent calculation overflow, caching mechanisms have to be implemented. E.g. in variable tab I cache evaluated formulas in a LRU-Cache so texts don't have to be parsed on every char entered. 

Also, what if user uses this button and takes over a just-typed variable value from waypoint notes, but then doesn't save the waypoint? Should then the taking-over get reverted?

--> very complex imho

**Comment by fm-sys** [comment]
BTW, the popup text should be shown as text content instead of title as it will be truncated otherwise:

![Screenshot_20230116_094251_cgeo geocaching](https://user-images.githubusercontent.com/64581222/212635050-a05772f3-0a9e-40ff-b65f-b66b3fb12d83.png)
